File: decode.py
import logging
from pathlib import Path
import sys


class Decoder:

    def __init__(self, file_path: Path) -> None:
        self.possible_solutions: list[dict] = []
        self.known_rules: dict = {}
        self.valid_solutions: list[dict] = []

        with open(file_path, "r") as f:
            self.contents: str = f.read()

        self.generations = self.contents.split()

    def decode(self) -> list[dict]:
        self.rules = {}
        self.axiom = self.generations[0]
        found_solution = False
        for i in range(1, len(self.generations)):
            """
            Each generation try to solve for the next generation with the current solutions
            If it is correct then keep going
            If it is not then through it away
            If none is found then do a full search by splitting the next generation
            """
            if not found_solution:
                logging.info(f"No found solution yet, decoding generation {i + 1}")
                local_solutions = self.find_possible_solutions(
                    self.generations[i - 1], self.generations[i]
                )

                self.update_possible_solutions(local_solutions)
                if len(self.possible_solutions) == 0:
                    logging.error("Program failed, no solutions found")
                    return []
                logging.debug(f"Number of possible solutions for generation {i + 1}: {len(self.possible_solutions)}")
                logging.debug(f"Known rules: {self.known_rules}")

                # iterate through the possible solutions to see if it solves the next ones
                solutions = self.possible_solutions.copy()
                for solution in solutions:
                    for j in range(1, len(self.generations)):
                        if (
                            self.generate(self.generations[j - 1], solution)
                            != self.generations[j]
                        ):
                            break

                        if j == (len(self.generations) - 1):
                            logging.info(f"Found solution: {solution}")
                            found_solution = True
                            if solution not in self.valid_solutions:
                                self.valid_solutions.append(solution)

        return self.valid_solutions

    def update_possible_solutions(self, new_solutions: list[dict]) -> None:
        previous_solutions = self.possible_solutions.copy()
        self.possible_solutions = []

        if len(new_solutions) == 1:
            self.known_rules.update(new_solutions[0])

        for new_sol in new_solutions:
            has_subset = False
            if self.is_conflicting_known_rules(new_sol):
                break
            for prev_sol in previous_solutions:
                if self.is_dict_subset(prev_sol, new_sol):
                    has_subset = True
                    prev_sol.update(new_sol)
                    self.possible_solutions.append(prev_sol)
            if not has_subset:
                self.possible_solutions.append(new_sol)

    def is_conflicting_known_rules(self, new_solution: dict) -> bool:
        for k_key, k_value in self.known_rules.items():
            if k_key in new_solution.keys():
                if new_solution[k_key] != k_value:
                    return True
        return False

    @staticmethod
    def is_dict_subset(dict_a: dict, dict_b: dict) -> bool:
        for key_a, value_a in dict_a.items():
            if key_a in dict_b.keys():
                if dict_b[key_a] != value_a:
                    return False
        return True

    def find_possible_solutions(
        self, previous_generation: str, current_generation: str
    ) -> list[dict]:
        splits = self.split_generation(
            previous_generation, current_generation, self.known_rules
        )
        return [dict(zip(previous_generation, split)) for split in splits]

    @staticmethod
    def generate(current_population: str, rules: dict) -> str:
        """
        Generate the next generation given the previous generation and a set of rules
        """
        new_population = ""
        for i in current_population:
            if i in rules.keys():
                new_population += rules[i]
            else:
                new_population += i

        return new_population

    @staticmethod
    def split_generation(
        previous_generation: str, current_generation: str, known_rules: dict
    ) -> list:
        def is_valid_partial_split(part, char, char_to_part):
            """
            Checks if the current split already appears in the strings and if so that they match
            """
            if char in char_to_part:
                return char_to_part[char] == part
            return True

        def split_recursive(
            A: str, B: str, current_split: list, char_to_part: dict, all_splits: list
        ) -> None:
            """
            Recursive function split B with the following rules
            1. Split the current generation for each corresponding character in the previous generation
            2. For each same character in the previous generation,
               the cooresponding groupd in the current generation must also match
            """
            if not A:
                if not B:
                    all_splits.append(current_split)
                return

            char = A[0]
            for i in range(1, len(B) - len(A) + 2):
                part = B[:i]
                if is_valid_partial_split(part, char, char_to_part):
                    new_char_to_part = char_to_part.copy()
                    new_char_to_part[char] = part
                    split_recursive(
                        A[1:],
                        B[i:],
                        current_split + [part],
                        new_char_to_part,
                        all_splits,
                    )

        all_splits = []
        split_recursive(
            previous_generation, current_generation, [], known_rules, all_splits
        )

        return all_splits


if __name__ == "__main__":
    logging.basicConfig(format="%(levelname)s: %(message)s", level=logging.DEBUG)
    sys.setrecursionlimit(100000)
    try:
        ruleset = Decoder(sys.argv[1]).decode()
    except RecursionError:
        print("Maximum recusion depth exceeded, try increasing the recursion limit")
        exit()

    if len(ruleset) < 1:
        print("Unable to find a solution, please provide more generations of input")
        exit()
    
    print("\n\nSolutions:")
    for rules in ruleset:
        for key, value in rules.items():
            print(f"{key} -> {value}")
        print("\n")

File: decode.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
#include <sstream>
#include <iterator>
#include <type_traits>

template <typename Key, typename Value>
class Decoder {
    static_assert(std::is_integral<Key>::value || std::is_same<Key, char>::value, "Key must be an integral type or char");

public:
    Decoder(const std::string& file_path) {
        std::ifstream file(file_path);
        if (file.is_open()) {
            std::ostringstream ss;
            ss << file.rdbuf();
            contents = ss.str();
            file.close();
        }

        std::istringstream iss(contents);
        generations = std::vector<std::string>((std::istream_iterator<std::string>(iss)),
                                                std::istream_iterator<std::string>());
    }

    std::vector<std::unordered_map<Key, Value> > decode() {
        rules.clear();
        axiom = generations[0];
        bool found_solution = false;
        for (size_t i = 1; i < generations.size(); ++i) {
            if (!found_solution) {
                std::cout << "No found solution yet, decoding generation " << i << std::endl;
                auto local_solutions = find_possible_solutions(generations[i - 1], generations[i]);

                update_possible_solutions(local_solutions);

                auto solutions = possible_solutions;
                for (const auto& solution : solutions) {
                    for (size_t j = 1; j < generations.size(); ++j) {
                        if (generate(generations[j - 1], solution) != generations[j]) {
                            break;
                        }

                        if (j == (generations.size() - 1)) {
                            std::cout << "Found solution: {";
                            for (const auto& pair : solution) {
                                std::cout << pair.first << ": " << pair.second << ", ";
                            }
                            std::cout << "}" << std::endl;
                            found_solution = true;
                            if (std::find(valid_solutions.begin(), valid_solutions.end(), solution) == valid_solutions.end()) {
                                valid_solutions.push_back(solution);
                            }
                        }
                    }
                }
            }
        }
        return valid_solutions;
    }

private:
    std::string contents;
    std::vector<std::string> generations;
    std::unordered_map<Key, Value> rules;
    std::string axiom;
    std::vector<std::unordered_map<Key, Value> > possible_solutions;
    std::vector<std::unordered_map<Key, Value> > valid_solutions;

    void update_possible_solutions(const std::vector<std::unordered_map<Key, Value> >& new_solutions) {
        auto previous_solutions = possible_solutions;
        possible_solutions.clear();

        for (const auto& new_sol : new_solutions) {
            bool has_subset = false;
            for (auto& prev_sol : previous_solutions) {
                if (is_dict_subset(prev_sol, new_sol)) {
                    has_subset = true;
                    for (const auto& pair : new_sol) {
                        prev_sol[pair.first] = pair.second;
                    }
                    possible_solutions.push_back(prev_sol);
                }
            }
            if (!has_subset) {
                possible_solutions.push_back(new_sol);
            }
        }
    }

    bool is_dict_subset(const std::unordered_map<Key, Value>& dict_a,
                        const std::unordered_map<Key, Value>& dict_b) const {
        for (const auto& [key_a, value_a] : dict_a) {
            auto it = dict_b.find(key_a);
            if (it != dict_b.end() && it->second != value_a) {
                return false;
            }
        }
        return true;
    }

    std::vector<std::unordered_map<Key, Value> > find_possible_solutions(const std::string& previous_generation,
                                                                        const std::string& current_generation) {
        auto splits = split_generation(previous_generation, current_generation);
        std::vector<std::unordered_map<Key, Value> > solutions;
        for (const auto& split : splits) {
            std::unordered_map<Key, Value> solution;
            for (size_t i = 0; i < previous_generation.size(); ++i) {
                solution[previous_generation[i]] = split[i];
            }
            solutions.push_back(solution);
        }
        return solutions;
    }

    std::string generate(const std::string& current_population,
                         const std::unordered_map<Key, Value>& rules) const {
        std::string new_population;
        for (char c : current_population) {
            if (rules.find(c) != rules.end()) {
                new_population += rules.at(c);
            } else {
                new_population += c;
            }
        }
        return new_population;
    }

    std::vector<std::vector<std::string> > split_generation(const std::string& previous_generation,
                                                           const std::string& current_generation) const {
        std::vector<std::vector<std::string> > all_splits;
        split_recursive(previous_generation, current_generation, {}, {}, all_splits);
        return all_splits;
    }

    void split_recursive(const std::string& A, const std::string& B,
                         std::vector<std::string> current_split,
                         std::unordered_map<Key, Value> char_to_part,
                         std::vector<std::vector<std::string> >& all_splits) const {
        if (A.empty()) {
            if (B.empty()) {
                all_splits.push_back(current_split);
            }
            return;
        }

        char char_a = A[0];
        for (size_t i = 1; i <= B.size() - A.size() + 1; ++i) {
            std::string part = B.substr(0, i);
            if (is_valid_partial_split(part, char_a, char_to_part)) {
                auto new_char_to_part = char_to_part;
                new_char_to_part[char_a] = part;
                auto new_current_split = current_split;
                new_current_split.push_back(part);
                split_recursive(A.substr(1), B.substr(i), new_current_split, new_char_to_part, all_splits);
            }
        }
    }

    bool is_valid_partial_split(const std::string& part, char char_a,
                                const std::unordered_map<Key, Value>& char_to_part) const {
        auto it = char_to_part.find(char_a);
        if (it != char_to_part.end()) {
            return it->second == part;
        }
        return true;
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <file_path>" << std::endl;
        return 1;
    }

    Decoder<char, std::string> decoder(argv[1]);
    auto rules = decoder.decode();
    std::cout << "Rules:" << std::endl;
    for (const auto& ruleset : rules) {
        for (const auto& pair : ruleset) {
            std::cout << pair.first << "->" << pair.second << std::endl;
        }
        std::cout << std::endl;
    }

    return 0;
}
